src: track cppgc wrappers with a list in Realm

joyeecheung opened the pull request:
src: track cppgc wrappers with a list in Realm

This allows us to perform cleanups of cppgc wrappers that rely
on a living Realm during Realm shutdown. Otherwise
the cleanup may happen during object destruction, which can
be triggered by GC after Realm shutdown, leading to invalid
access to Realm.

The general pattern for this type of non-trivial destruction is
designed to be:

```c++
class MyWrap final : CPPGC_MIXIN(MyWrap) {
 public:
  ~MyWrap() { this->Finalize(); }
  void Clean(Realm* realm) override {
     // Do cleanup that relies on a living Realm. This would be
     // called by CppgcMixin::Finalize() first during Realm shutdown,
     // while the Realm is still alive. If the destructor calls
     // Finalize() again later during garbage collection that happens after
     // Realm shutdown, Clean() would be skipped, preventing
     // invalid access to the Realm.
  }
}
```

In addition, this allows us to trace external memory held by the wrappers
in the heap snapshots if we add synthethic edges between the wrappers
and other nodes in the embdder graph callback, or to perform snapshot
serialization for them.

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

legendecas commented on src/cppgc_helpers.h:
```diff
@@ -88,8 +99,29 @@ class CppgcMixin : public cppgc::GarbageCollectedMixin {
+  // destructor. Outside of CleanEnvResource(), subclasses should avoid calling
+  // into JavaScript or perform any operation that can trigger garbage
+  // collection during the destruction.
+  void Clean() override {
```
Looks like a good pattern of `Cleanable`. Can this be merged into the `Environment::cleanable_queue_`?

https://github.com/nodejs/node/blob/b8f6d84ff5893b1274d658da22bc4067c2fb93fa/src/env.h#L606-L616

joyeecheung replied:
That was what I did at first :) but then I realized if it's in that queue, to filter out all the wraps it would be an unnecessary hassle. Might as well just keep them in a separate queue and there would be no need for filtering.

legendecas commented on src/README.md:
````diff
@@ -1206,6 +1219,27 @@ it, this can happen at any time after the garbage collector notices that
+the `Environment` is already gone, it must implement the cleanup with this pattern:
+
+```c++
+  ~MyWrap() { this->Clean(); }
````
The destructor can be invoked by the garbage collector, and `Clean()` will access heap objects. The garbage collector can invoke the destructor at non-creation thread. This example should depend on `CPPGC_USING_PRE_FINALIZER`.

joyeecheung replied:
Updated to mention `CPPGC_USING_PRE_FINALIZER`, though I noticed two things:

1. `ContextifyContext` does not even need to have special cleanup and can just have a default destructor - when the destructors are invoked, the context is already going away as it would've been holding the ContextifyContext alive. So there's no need to set the internal pointers of the context; On the other hand since the context is already going away, the per-environment context list would now contain empty global handles to that context, and the context tracking code is already able to deal with empty handles, so it matters little to purge it from the list in the destructor (we could also do the housekeeping in the constructor, in `TrackContext()` instead, to make sure there aren't too many empty handles lying around).
2. It seems the use of `CPPGC_USING_PRE_FINALIZER` would crash the construction of `ContextifyContext`, which I have not gotten to the bottom of yet.

legendecas commented on src/README.md:
````diff
@@ -1206,6 +1228,51 @@ it, this can happen at any time after the garbage collector notices that
+   pattern in its class body:
+
+   ```cpp
+   ~MyWrap() { this->Clean(); }
+   void CleanEnvResource(Environment* env) override {
+     // Do cleanup that relies on a living Environemnt. This would be
+     // called by CppgcMixin::Clean() first during Environment shutdown,
+     // while the Environment is still alive. If the destructor calls
+     // Clean() again later during garbage collection that happens after
+     // Environment shutdown, CleanEnvResource() would be skipped, preventing
+     // invalid access to the Environment.
+   }
````
I'm still in doubt if this example will entice people to access handles on the `node::Environment` given that it is passed in as a parameter, and ultimately call this `CleanEnvResource` through the destructor, which is invoked by cppgc.

I'm wondering if we should avoid creating a cppgc object that is short-lived than its associated env resources.

joyeecheung replied:
Isn't the example in 3 enough to address this?  I can remove `Environment` as a parameter though I don't think it makes much of a difference - people can still easily reach out to `env()` if they want, just like in the case of BaseObject's destructors, even if there destructors are called as the first pass weak callback.

legendecas replied:
Right, I think my comment was purely hypothetical.

jasnell commented on src/README.md:
```diff
@@ -1206,6 +1228,51 @@ it, this can happen at any time after the garbage collector notices that
+the cleanups of a cppgc-managed object should adhere to different patterns,
+depending on what it needs to do:
+
+1. If it does not need to do any non-trivial, nor does its members, just use
```
sentence appears to be incomplete? Do any non-trivial what?

jasnell commented on src/README.md:
```diff
@@ -1206,6 +1228,51 @@ it, this can happen at any time after the garbage collector notices that
+1. If it does not need to do any non-trivial, nor does its members, just use
+   the default destructor.
+2. If the cleanup relies on a living `Environment`, the class should use this
+   pattern in its class body:
```
This term "living `Environment`" is not defined. What is a living environment?

legendecas replied:
In this context, the term "living Enviroment" refers to a condition where a cppgc object can out-live its `Environment` as the cppgc heap is cleaned up after the `Environment` been destructed, as documented above:

> When a cppgc-managed object is no longer reachable in the heap, its destructor
will be invoked by the garbage collection, which can happen after the `Environment`
is already gone.

Though, it didn't mention the term "living Environment".

jasnell replied:
Yeah, I know what was meant, but it should be explained a bit more in the document so that it is clear.

joyeecheung commented:
Trying to rebase https://github.com/nodejs/node/pull/51017 on top of it to verify the "use list for heap snapshot integration" idea. It seems we'll need to update the MemoryTracker code to stop creating duplicate nodes for cppgc wrappers. Still looking into it..

jasnell commented on src/cppgc_helpers.h:
```diff
@@ -48,49 +61,55 @@ class CppgcMixin : public cppgc::GarbageCollectedMixin {
+  // collection during the destruction.
+  void Clean() {
+    if (realm_ == nullptr) return;
+    this->CleanEnvResource(realm_);
```
Super minor but... Should this include any reentrancy protections? If the `CleanEnvResource(...)` impl somehow ends up calling `Clean()` again.

joyeecheung replied:
I think it only depends on whether the `CleanEnvResource` implemented is re-enterable or intended to be re-entered. Both might be deseriable and it's fully up to the implementer.

jasnell replied:
Ok. It *might* be worth at least mentioning in the docs or comments for `CleanEnvResource(...)` that it is up to the implementer to handle re-entrancy, but definitely minor and non-blocking.

joyeecheung commented on src/README.md:
````diff
@@ -1266,16 +1277,60 @@ referrer->Set(
+   ```cpp
+    private:
+     CPPGC_USING_PRE_FINALIZER(MyWrap, Clean);
+   ```
````
> let's say someone has a wrapped object that is holding a `BaseObjectPtr<Foo>` as a member

I don't think this would be supported in the nearest future, or will ever be supported - a cppgc-object should not hold on to a `BaseObjectPtr<Foo>` without converting that `BaseObjectPtr<Foo>` to cppgc-managed first - and then in that case they should use `cppgc::Member<Foo>` to hold on to that instead. The migration needs to be done bottom-to-top. In the case of using `cppgc::Member<Foo>`, nothing special needs to be done.

> What about if it has a v8::Global<...> member field?

I think in most cases, this should be `v8::TracedReference` instead. In that case nothing special needs to be done in the destructor.

jasnell replied:
I wonder if we can somehow have c++ linting detect it for us if someone includes a `BaseObjectPtr` member in one of these? .... in any case, this all sounds fine, I'd just like to make sure this gets included somehow in the documentation.

joyeecheung replied:
If it comes up in the future that some people attempt to use `BaseObjectPtr` in a cppgc wrapper, we can consider adding some linting rules. For now it seems a bit too premature. Since we are still in an early phase of cppgc migration, I think the rule of thumb is probably: 

- If you are not looking into also updating the cppgc helpers, only use cppgc on classes that aren't more than what existing cppgc wrappers are doing (at this point in time, v8 handles and MemoryRetainer over other plain types are supported, cppgc members are supposed to be supported by we haven't tried it internally yet)
- If you want to tread on untrodden paths and manage more complex memory structures, like a cppgc wrapper referencing BaseObject, expect to figure out how to make it work and update the cppgc helpers yourself.

joyeecheung commented:
Added more docs and tests. CI is having some memory issue on Windows build machine. Apart from that things look good. I updated OP with the follow-up plan but I think this should be fine to land as an initial implementation. Can I have another look please?  @jasnell @mcollina

joyeecheung commented:
Adding the realm method to make the debug build happy..

joyeecheung commented:
The crash in the CI looks relevant, but then when I tried running the test locally on Windows, it seems the main branch is crashing at memory issues as well? Still trying to figure out what's going on..

joyeecheung commented:
Tracking the cppgc wrappers with a list of weak persistents instead seems to make the crash go away. Trying a CI..

joyeecheung commented:
It seems the windows arm64 crash is going with the V8 13.6 upgrade. Trying again..

joyeecheung commented:
Rebased after https://github.com/nodejs/node/pull/58190 landed to remove the flake..

joyeecheung commented:
CI is happy. @jasnell @legendecas can you take another look please? The last commit updated the list to hold weak persistents and uses the GC epilogue callback to purge empty weak persistents - mutating the handles at destruction incurs more complexity than it needs - letting V8's GC marking them as empty once the referenced objects become unreachable so that we can recognize and skip is already good enough for our own iterations. We just need to regularly prune the useless handles from list.

joyeecheung commented:
CI is struggling again due to https://github.com/nodejs/node/issues/57235...

aduh95 commented:
This does not land cleanly on v22.x-staging and would require a manual backport if we want it on v22.x
